### PR TITLE
CORS for API access via service accounts and OAuth

### LINKED ISSUE #961: CORS Settings Service Accounts (6h)
Bei CEVI und SBV soll die CORS-Policy: Acces-Control-Allow-Origin aktiviert werden.

Da ein potentieller Angreifer mit CORS und der USER-API in Zusammenhang mit dem entsprechenden Cookie zuviel Daten auslesen könnte, muss diese im selben Zug deaktiviert werden.

## Techspec

- [ ] AllowedOrigins Feld zu ServiceToken hinzufügen
- [ ] rack-cors Gem verwenden
- [ ] Authenticatable::normalize_user_params in eine Middleware-kompatible Domain-Klasse auslagern
- [ ] Domain-Klasse in rack-cors initializer verwenden (origins mit Block) um Token zu finden und nach erlaubten Origins zu befragen

## Original Story

Ich habe eben mit einer JavaScript-Anwendung zur Darstellung der Kursinfos gespielt und scheitere dabei an folgender Fehlermeldung:
```
Access to XMLHttpRequest at 'https://db.cevi.ch/[...]' from origin '[...]' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```
Wenn ich das richtig verstehe fehlt dieser Header auf Seiten von Hitobito und ohne können die Service Accounts nicht via JavaScript genutzt werden. Im Moment wäre wohl ein freier Access ok. Längerfristig wäre es aber wohl sinnvoller, wenn man die erlaubten Origin beim API-Key definieren könnte.

### PR DESCRIPTION
**Pull request by carlobeltrame**
Fixes #961. Created during the hitobito community hackathon 2021 :partying_face:

**tl;dr:** This PR allows to use the JSON API using JavaScript on whitelisted websites. Previously, the API could only be consumed server-side, but not in a user's browser.

This adds the [rack-cors](https://github.com/cyu/rack-cors) gem and configures its responses dynamically based on allowed CORS hosts of the used OAuth access_token or service token. In order to not repeat any existing logic, I also refactored authenticatable.rb and extracted some reusable code from it into a domain class `Authenticatable::Tokens` (and removed the old and hacky `normalize_user_params` method). This class is used in authenticatable.rb and in the new `Api::CorsCheck` domain class, which is responsible for telling rack-cors whether a request origin is allowed or not. Overall, I hope the authentication logic is a little clearer to read now as well.

@diegosteiner @MartinGantenbein @Michael-Schaer if you have time, you might want to verify whether your existing OAuth and OIDC applications / service token API clients still work after this change. I have some of my own client applications, but the more testing is done the better.

To do:
- [ ] Manually test that the user tokens, service tokens and OAuth API tokens all still work as documented
- [ ] Write specs for the new domain classes

### CODE REVIEW

**Comment by MartinGantenbein** [comment]
Still works for my application, thanks for alerting me.

File: doc/development/05_rest_api.md [hunk 1]
```diff
@@ -121,7 +121,6 @@ Currently the following endpoints are provided:
 
 | Method | Path                                            | Function                                                                        |
 | ---    | ---                                             | ---                                                                             |
-| GET    | /groups                                         | Redirects to root group (only works using personal token, not service accounts) |
```
Comment (MartinGantenbein) [inline doc/development/05_rest_api.md:124]: Was /groups really removed?

Comment (carlobeltrame) [reply doc/development/05_rest_api.md:124]: Yes, I removed the endpoint `/groups` in 7794588fc656b6c7acf2f016d9c6aa301d446385 because
1. it just redirected to the path of the root group, but redirects in JSON APIs generally require more effort from the client to follow correctly
1. it only worked when using personal user tokens, it didn't work with service accounts or OAuth API access
1. API clients are mostly programmed for a fixed hitobito instance, and the root group id can easily be configured on the client if needed
1. it's easy to add back in a more consistent way once this PR is merged, if there really is demand for this endpoint

Meanwhile, retrieving a single group using the `/groups/{id}` endpoint still works of course.

Is there a need for this specific endpoint in your application? What are your thoughts?

Comment (MartinGantenbein) [reply doc/development/05_rest_api.md:124]: Thanks for the context, that's fine with me.

**Comment by carlobeltrame** [comment]
Push @nchiapol

**Comment by nchiapol** [comment]
@carlobeltrame Mir ist dein Kommentar nicht ganz klar. Muss/Soll ich etwas unternehmen?

**Comment by carlobeltrame** [comment]
Nein, habe dich nur markiert weil euch dieser PR gemäss Community Meeting betrifft, mehr zur Info.

**Comment by nchiapol** [comment]
Alles klar, danke. Werde das gerne ausprobieren, sobald es gemerged ist.

File: Gemfile.lock [hunk 2]
```diff
@@ -391,6 +391,8 @@ GEM
     puma (4.3.5)
       nio4r (~> 2.0)
     rack (2.2.3)
+    rack-cors (1.1.1)
```
Comment (psunix) [inline Gemfile.lock:394]: version 1.1.1 ist vom Dezember 2019.
jedoch gibt es im github repo des projekts commits vom Dezember 2020. scheint also noch aktiv zu sein.

**Comment by betsim** [comment]
Wir warten schon sehnlichst auf dieses Feature 🥳

**Comment by carlobeltrame** [comment]
@betsim das Feature ist jetzt live bei euch :+1:
